File: src/lib/commandGateway.ts
/**
 * MIA by VIA — Command Gateway & State Orchestration Kernel
 * Pipeline: Command -> Validate Schema -> Authority Check -> Current State -> Policy Evaluation -> Execute Transition -> Emit Event -> Seal Ops Receipt
 */

import {
  CITIZEN_IDENTITY_TRANSITIONS,
  CREDENTIAL_TRANSITIONS,
  GCODE_TRANSITIONS,
  PROOF_REQUEST_TRANSITIONS,
  VALUE_INSTRUCTION_TRANSITIONS,
  type CitizenIdentityState,
  type CitizenIdentityTrigger,
  type CredentialState,
  type CredentialTrigger,
  type GovernmentCodeState,
  type GovernmentCodeTrigger,
  type ProofRequestState,
  type ProofRequestTrigger,
  type ValueInstructionState,
  type ValueInstructionTrigger
} from './stateMachines';

export interface Command {
  commandId: string;
  aggregateType: 'CitizenIdentity' | 'CitizenNode' | 'GovernmentCode' | 'Credential' | 'ProofRequest' | 'AccessGrant' | 'ServiceCase' | 'CivicValueAccount' | 'ValueInstruction';
  aggregateId: string;
  trigger: string;
  actorRole: 'Citizen' | 'GCodeOwner' | 'PlatformOperator' | 'GovernanceAuthority';
  actorId: string;
  payload: Record<string, any>;
  timestamp: string;
}

export interface TransitionResult {
  success: boolean;
  commandId: string;
  aggregateType: string;
  aggregateId: string;
  previousState: string;
  newState: string;
  eventHash: string;
  receiptId: string;
  policyDecision: string;
  error?: string;
}

class CommandGatewayKernel {
  private aggregateStates: Record<string, string> = {
    'id-citizen-9921': 'ACTIVE',
    'node-cnode-9921': 'ACTIVE',
    'did:via:gcode:permit-building-01': 'ACTIVE',
    'vc-solar-8841': 'ACTIVE',
    'prf-req-101': 'CREATED',
    'val-inst-901': 'DRAFT'
  };

  private receiptsChain: Array<{
    receiptId: string;
    commandId: string;
    aggregateId: string;
    prevState: string;
    newState: string;
    prevHash: string;
    receiptHash: string;
    timestamp: string;
  }> = [];

  constructor() {
    // Genesis receipt
    const genesisHash = this.sha256('GENESIS_RECEIPT_G0_INTENT');
    this.receiptsChain.push({
      receiptId: 'rcpt-000',
      commandId: 'cmd-genesis',
      aggregateId: 'system-kernel',
      prevState: 'NULL',
      newState: 'INITIALIZED',
      prevHash: '0000000000000000000000000000000000000000000000000000000000000000',
      receiptHash: genesisHash,
      timestamp: new Date().toISOString()
    });
  }

  private sha256(input: string): string {
    // Deterministic string hash for demonstration receipt sealing
    let hash = 0;
    for (let i = 0; i < input.length; i++) {
      const char = input.charCodeAt(i);
      hash = (hash << 5) - hash + char;
      hash |= 0;
    }
    const hex = Math.abs(hash).toString(16).padStart(8, '0');
    return '0x' + hex.repeat(8).substring(0, 64);
  }

  public executeCommand(cmd: Command): TransitionResult {
    const currentState = this.aggregateStates[cmd.aggregateId] || 'DRAFT';

    // 1. Authority Guard Check (Separation of Duties)
    if (cmd.actorRole === 'Citizen' && ['AUTHORIZE_GCODE', 'REVOKE_CREDENTIAL'].includes(cmd.trigger)) {
      return {
        success: false,
        commandId: cmd.commandId,
        aggregateType: cmd.aggregateType,
        aggregateId: cmd.aggregateId,
        previousState: currentState,
        newState: currentState,
        eventHash: '',
        receiptId: '',
        policyDecision: 'DENIED_AUTHORITY_VIOLATION',
        error: `Actor ${cmd.actorId} with role ${cmd.actorRole} cannot execute ${cmd.trigger}`
      };
    }

    // 2. Validate Allowed Transition Matrix
    let newState: string | undefined = undefined;

    if (cmd.aggregateType === 'CitizenIdentity') {
      newState = CITIZEN_IDENTITY_TRANSITIONS[currentState as CitizenIdentityState]?.[cmd.trigger as CitizenIdentityTrigger];
    } else if (cmd.aggregateType === 'Credential') {
      newState = CREDENTIAL_TRANSITIONS[currentState as CredentialState]?.[cmd.trigger as CredentialTrigger];
    } else if (cmd.aggregateType === 'GovernmentCode') {
      newState = GCODE_TRANSITIONS[currentState as GovernmentCodeState]?.[cmd.trigger as GovernmentCodeTrigger];
    } else if (cmd.aggregateType === 'ProofRequest') {
      newState = PROOF_REQUEST_TRANSITIONS[currentState as ProofRequestState]?.[cmd.trigger as ProofRequestTrigger];
    } else if (cmd.aggregateType === 'ValueInstruction') {
      newState = VALUE_INSTRUCTION_TRANSITIONS[currentState as ValueInstructionState]?.[cmd.trigger as ValueInstructionTrigger];
    } else {
      newState = 'ACTIVE';
    }

    if (!newState) {
      return {
        success: false,
        commandId: cmd.commandId,
        aggregateType: cmd.aggregateType,
        aggregateId: cmd.aggregateId,
        previousState: currentState,
        newState: currentState,
        eventHash: '',
        receiptId: '',
        policyDecision: 'INVALID_TRANSITION',
        error: `Transition ${cmd.trigger} not allowed from state ${currentState}`
      };
    }

    // 3. Update State
    this.aggregateStates[cmd.aggregateId] = newState;

    // 4. Generate Hash-Chained Receipt
    const prevReceipt = this.receiptsChain[this.receiptsChain.length - 1];
    const receiptContent = `${prevReceipt.receiptHash}:${cmd.commandId}:${cmd.aggregateId}:${currentState}:${newState}:${cmd.timestamp}`;
    const receiptHash = this.sha256(receiptContent);

    const receipt = {
      receiptId: `rcpt-${Date.now()}`,
      commandId: cmd.commandId,
      aggregateId: cmd.aggregateId,
      prevState: currentState,
      newState: newState,
      prevHash: prevReceipt.receiptHash,
      receiptHash: receiptHash,
      timestamp: cmd.timestamp
    };

    this.receiptsChain.push(receipt);

    return {
      success: true,
      commandId: cmd.commandId,
      aggregateType: cmd.aggregateType,
      aggregateId: cmd.aggregateId,
      previousState: currentState,
      newState: newState,
      eventHash: this.sha256(JSON.stringify(cmd)),
      receiptId: receipt.receiptId,
      policyDecision: 'POLICY_PASSED_SEALED'
    };
  }

  public getReceipts() {
    return this.receiptsChain;
  }

  public getAggregateState(id: string) {
    return this.aggregateStates[id] || 'DRAFT';
  }
}

export const commandGateway = new CommandGatewayKernel();

File: rust/src/command_gateway.rs
//! Command Gateway & State Orchestrator Kernel
//! Enforces: Command -> Validate Schema -> Authority Check -> Current State -> Policy Evaluation -> Execute Transition -> Emit Event -> Seal Receipt

use crate::receipt_sealer::EvidenceReceiptSealer;
use crate::state_machines::{
    CitizenIdentityState, CitizenIdentityTrigger, CredentialState, CredentialTrigger,
    GovernmentCodeState,
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ActorRole {
    Citizen,
    GCodeOwner,
    PlatformOperator,
    GovernanceAuthority,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Command {
    pub command_id: String,
    pub aggregate_type: String,
    pub aggregate_id: String,
    pub trigger: String,
    pub actor_role: ActorRole,
    pub actor_id: String,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TransitionExecutionResult {
    pub success: bool,
    pub command_id: String,
    pub aggregate_id: String,
    pub previous_state: String,
    pub new_state: String,
    pub receipt_hash: String,
    pub policy_decision: String,
    pub error: Option<String>,
}

pub struct CommandGatewayKernel {
    pub citizen_identity_states: HashMap<String, CitizenIdentityState>,
    pub credential_states: HashMap<String, CredentialState>,
    pub gcode_states: HashMap<String, GovernmentCodeState>,
    pub receipt_sealer: EvidenceReceiptSealer,
}

impl CommandGatewayKernel {
    pub fn new() -> Self {
        let mut gateway = Self {
            citizen_identity_states: HashMap::new(),
            credential_states: HashMap::new(),
            gcode_states: HashMap::new(),
            receipt_sealer: EvidenceReceiptSealer::new(),
        };

        // Seed default states
        gateway.citizen_identity_states.insert("id-citizen-9921".to_string(), CitizenIdentityState::Draft);
        gateway.credential_states.insert("vc-solar-8841".to_string(), CredentialState::Issued);
        gateway.gcode_states.insert("did:via:gcode:permit-building-01".to_string(), GovernmentCodeState::Authorized);

        gateway
    }

    pub fn execute_command(&mut self, cmd: Command) -> TransitionExecutionResult {
        // 1. Separation of Duties Check
        if cmd.actor_role == ActorRole::Citizen && (cmd.trigger == "AUTHORIZE_GCODE" || cmd.trigger == "REVOKE_CREDENTIAL") {
            return TransitionExecutionResult {
                success: false,
                command_id: cmd.command_id,
                aggregate_id: cmd.aggregate_id,
                previous_state: "UNKNOWN".to_string(),
                new_state: "UNKNOWN".to_string(),
                receipt_hash: "".to_string(),
                policy_decision: "DENIED_AUTHORITY_VIOLATION".to_string(),
                error: Some(format!("Role {:?} unauthorized for trigger {}", cmd.actor_role, cmd.trigger)),
            };
        }

        // 2. Execute Aggregate State Transition
        if cmd.aggregate_type == "CitizenIdentity" {
            let current = self
                .citizen_identity_states
                .get(&cmd.aggregate_id)
                .cloned()
                .unwrap_or(CitizenIdentityState::Draft);

            let trigger = match cmd.trigger.as_str() {
                "SUBMIT_ENROLLMENT" => CitizenIdentityTrigger::SubmitEnrollment,
                "APPROVE_VERIFICATION" => CitizenIdentityTrigger::ApproveVerification,
                "SUSPEND_IDENTITY" => CitizenIdentityTrigger::SuspendIdentity,
                _ => return self.error_result(&cmd, &current.to_string(), "Unsupported trigger"),
            };

            match current.transition(trigger) {
                Ok(next_state) => {
                    self.citizen_identity_states.insert(cmd.aggregate_id.clone(), next_state);
                    let rcpt = self.receipt_sealer.seal_receipt(
                        "CITIZEN_IDENTITY_TRANSITION",
                        &cmd.command_id,
                        &cmd.aggregate_id,
                        &current.to_string(),
                        &next_state.to_string(),
                        vec![cmd.actor_id.clone()],
                    );

                    TransitionExecutionResult {
                        success: true,
                        command_id: cmd.command_id,
                        aggregate_id: cmd.aggregate_id,
                        previous_state: current.to_string(),
                        new_state: next_state.to_string(),
                        receipt_hash: rcpt.receipt_hash,
                        policy_decision: "PASSED_SEALED".to_string(),
                        error: None,
                    }
                }
                Err(e) => self.error_result(&cmd, &current.to_string(), e),
            }
        } else if cmd.aggregate_type == "Credential" {
            let current = self
                .credential_states
                .get(&cmd.aggregate_id)
                .cloned()
                .unwrap_or(CredentialState::Draft);

            let trigger = match cmd.trigger.as_str() {
                "ACTIVATE_CREDENTIAL" => CredentialTrigger::ActivateCredential,
                "SUSPEND_CREDENTIAL" => CredentialTrigger::SuspendCredential,
                "REVOKE_CREDENTIAL" => CredentialTrigger::RevokeCredential,
                _ => return self.error_result(&cmd, &current.to_string(), "Unsupported trigger"),
            };

            match current.transition(trigger) {
                Ok(next_state) => {
                    self.credential_states.insert(cmd.aggregate_id.clone(), next_state);
                    let rcpt = self.receipt_sealer.seal_receipt(
                        "CREDENTIAL_TRANSITION",
                        &cmd.command_id,
                        &cmd.aggregate_id,
                        &current.to_string(),
                        &next_state.to_string(),
                        vec![cmd.actor_id.clone()],
                    );

                    TransitionExecutionResult {
                        success: true,
                        command_id: cmd.command_id,
                        aggregate_id: cmd.aggregate_id,
                        previous_state: current.to_string(),
                        new_state: next_state.to_string(),
                        receipt_hash: rcpt.receipt_hash,
                        policy_decision: "PASSED_SEALED".to_string(),
                        error: None,
                    }
                }
                Err(e) => self.error_result(&cmd, &current.to_string(), e),
            }
        } else {
            self.error_result(&cmd, "DRAFT", "Aggregate type pending implementation")
        }
    }

    fn error_result(&self, cmd: &Command, prev_state: &str, err_msg: &str) -> TransitionExecutionResult {
        TransitionExecutionResult {
            success: false,
            command_id: cmd.command_id.clone(),
            aggregate_id: cmd.aggregate_id.clone(),
            previous_state: prev_state.to_string(),
            new_state: prev_state.to_string(),
            receipt_hash: "".to_string(),
            policy_decision: "INVALID_TRANSITION".to_string(),
            error: Some(err_msg.to_string()),
        }
    }
}
